### DISCUSSION TITLE
Support for Debian on Raspberry Pi 4B/CM4?

### DISCUSSION THREAD

**Discussion by HankB** [zbm-dev/zfsbootmenu#464] [General]
Is there any chance for something along these lines? Current status of ZFS on Raspberry Pi:

* R-Pi OS - seemingly not a priority for the R-Pi engineers. They frequently break ZFS on existing installs. I wouldn;t pursue this.
* Ubuntu 22.04 - Bootable with ZFS on root (with instructions at <https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS%20for%20Raspberry%20Pi.html>. Instructions mention the possibility of using U-boot.
* I've been working unsuccessfully to get Debian working with ZFS on root by using as much of the Ubuntu instructions as seem likely.
* Report of Gentoo working with ZFS on root. <https://forums.raspberrypi.com/viewtopic.php?t=348170>

I can find no mention of using ZFS Boot menu on a Raspberry Pi. I did find some information about using boot environments on a Pi running OpenBSD.

My preference would be Debian and I wonder of ZFS boot menu might get me over the hump. Ubuntu could be an option. I'm not particularly interested in using Gentoo or OpenBSD.

The closest I came to getting Debian with ZFS on root was to get everything except `/` on ZFS. I could go that way and perhaps tailor it so that most things wound up on ZFS. Some of my notes on this are at <https://github.com/HankB/MkDocs-blog/blob/main/docs/tech/Pi-Debian-Root-on-ZFS.md>

Thanks!

**Comment by classabbyamp** [comment]
if you can get `kexec` to work on an RPi, then sure it should be possible

**Comment by ahesford** [comment]
I have long been (passively) interested in getting ZFSBootMenu working on a Raspberry Pi, but I have not been successful in making `kexec` work. Instead, I just keep `/boot` as a FAT32 volume that contains the kernel, standard RPi bootloader files and an initramfs image. (The RPi bootloader can be told to load an initramfs image by [adding an entry to config.txt](https://www.raspberrypi.com/documentation/computers/config_txt.html#initramfs); you should be able to configure your package manager to generate a suitable initramfs image that includes ZFS drivers after every kernel upgrade.)

**Reply by HankB** [reply]
Thanks for the reply.

The FAT32 volume is (AFAIK) a hard requirement due to how the Pi boots. I'm OK with that. I've worked with initramfs entries in `/boot/firmware`. Here's where I'm at.

* the initrd appears to have the zfs module included (edit: along with a lot of other ZFS related stuff.)

```text
root@mars:/boot/firmware# lsinitramfs initrd.img-6.1.0-11-arm64|grep zfs.ko
usr/lib/modules/6.1.0-11-arm64/updates/dkms/zfs.ko
root@mars:/boot/firmware# 
```

Kernel and initrd listed in `config.txt`

```text
root@mars:/boot/firmware# grep 6.1.0.11 config.txt 
kernel=vmlinux-6.1.0-11-arm64
initramfs initrd.img-6.1.0-11-arm64 followkernel
root@mars:/boot/firmware# 
```

I just added  `followkernel` to the `initramfs` statement. When I run `update-initramfs` it produces a kernel named `vmlinuz-6.1.0-11-arm64` and my understanding is that this is supposed to denote a compressed kernel. It is not compressed as far as I can tell. I have tried renaming to `vmlinux-6.1.0-11-arm64` which seems to make no difference. The cmdline.cfg I'm using is

```text
console=tty0 console=ttyS1,115200 root=ZFS=rpool/ROOT/debian rw fsck.repair=yes net.ifnames=0  rootwait zfsforce=1 debug
```

The page for InitRamfs debugging <https://wiki.debian.org/InitramfsDebug> suggests that `debug` should produce a lot of output which it seems not to.  The system comes up to the point where the last line (initializing a USB keyboard) is

```text
[   7.892116] hid-generic ...
```

This is a bit after it recognizes the SSD (`/dev/sda`) and if I hit `<ctrl><alt><del>` It responds but cannot reboot.

Any suggestions for what I've overlooked are most welcome! 

(This is verging off topic but I appreciate the help. We can continue elsewhere if desirable.)

Edit: I cannot find the `followkernel` keyword anywhere in the Debian initramfs documentation. Is it an R-Pi OS extension?

**Reply by ahesford** [reply]
The file `config.txt` isn't parsed by the Linux kernel or by any initramfs modules, but by the built-in Raspberry Pi bootloader.

Your kernel probably isn't even getting far enough to launch `init` within the initramfs. The Raspberry Pi foundation maintains a fork of the Linux kernel that it builds for Raspberry Pi OS and, while support for the hardware is all supposed to be available in the upstream kernel, I'm not sure that a stock Debian aarch64 kernel configuration includes all of the devices necessary. There is also some device-tree magic that goes on when booting the RPi kernel.

**Reply by HankB** [reply]
Thanks for taking the time to write that. It did seem that the initramfs was not getting off the ground, for whatever reason. 

I did have a go at `kexec` which is in the repos but seems not to work. I did find one report of it working on Arch on a Pi 3B but lacked sufficient detail to give me any clue. I think I'll set this aside for the moment. I should retrace my steps that got me to "almost root on ZFS" and that can be useful for now.
